# Assignment 2: Linear Models and Validation Metrics (30 marks total)
### Due: October 10 at 11:59pm

### Name: Mevin Moncy

### In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.

## Part 1: Classification (14.5 marks total)

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:

### Step 0: Import Libraries

In [1]:
import numpy as np
import pandas as pd

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [2]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y


from yellowbrick.datasets import load_spam
import mglearn
import yellowbrick

X,y = load_spam()

# TO DO: Print size and type of X and y
print(X.shape)
print(y.shape)

print(type(X))
print(type(y))
print()

print(X.dtypes)
print(y.dtype)

(4600, 57)
(4600,)
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>

word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float64
word_freq_000                 float64
word_freq_money               fl

### Step 2: Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [3]:
# TO DO: Check if there are any missing values and fill them in if necessary

missing_values_X = X.isnull().sum().sum()
missing_values_y = y.isnull().sum().sum()

print(missing_values_X)
print(missing_values_y)

0
0


For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [4]:
# TO DO: Create X_small and y_small 
from sklearn.model_selection import train_test_split #splits training and testing sets, x IS DEATURE, Y IS TARGET, X_small,ymall is the training

X_small, X_val, y_small, y_val = train_test_split(X, y,train_size=0.05, random_state=0)
X_small_a, X_val_a, y_small_a, y_val_a = train_test_split(X_small, y_small, random_state=0)


### Step 3: Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets: 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

In [5]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=2000)

#Spliting into three datasets:

## X & y
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
model_logreg = LogisticRegression(max_iter=2000).fit(X_train, y_train)

# Only first two columns of X and y
X_train_partial, X_val_partial, y_train_partial, y_val_partial = train_test_split(X.iloc[:, :2], y, random_state=0)
model_partial = LogisticRegression(max_iter=2000).fit(X_train_partial, y_train_partial)

# X_small and y_small
X_small, X_val_small, y_small, y_val_small = train_test_split(X, y, train_size=0.05, random_state=0)
X_small_a, X_val_a, y_small_a, y_val_a = train_test_split(X_small, y_small, random_state=0)
model_small = LogisticRegression(max_iter=2000).fit(X_small_a, y_small_a)


### Step 4: Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

In [6]:
## X & y
model_logreg_training = model_logreg.score(X_train, y_train)
model_logreg_val = model_logreg.score(X_val, y_val)
print(f"Training set score for X & y: {model_logreg_training:.3f}")
print(f"Validation set score for X & y: {model_logreg_val:.3f}")
print()

# Only first two columns of X and y
model_partial_training = model_partial.score(X_train_partial, y_train_partial)
model_partial_val = model_partial.score(X_val_partial, y_val_partial)
print(f"Training set score for partial X & y: {model_partial_training:.3f}")
print(f"Validation set score for partial X & y: {model_partial_val:.3f}")
print()

# X_small and y_small
model_small_training = model_small.score(X_small_a, y_small_a)
model_small_val = model_small.score(X_val_a, y_val_a)
print(f"Training set score for small X & y: {model_small_training:.3f}")
print(f"Validation set score for small X & y: {model_small_val:.3f}")

Training set score for X & y: 0.928
Validation set score for X & y: 0.938

Training set score for partial X & y: 0.608
Validation set score for partial X & y: 0.613

Training set score for small X & y: 0.936
Validation set score for small X & y: 0.931


### Step 5: Visualize Results (4 marks)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [7]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT

#####------------Implement ML Model

from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=2000)

#Spliting into three datasets:

## X & y
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
model_logreg = LogisticRegression(max_iter=2000).fit(X_train, y_train)

# Only first two columns of X and y
X_train_partial, X_val_partial, y_train_partial, y_val_partial = train_test_split(X.iloc[:, :2], y, random_state=0)
model_partial = LogisticRegression(max_iter=2000).fit(X_train_partial, y_train_partial)

# X_small and y_small
X_small, X_val_small, y_small, y_val_small = train_test_split(X, y, train_size=0.05, random_state=0)
X_small_a, X_val_a, y_small_a, y_val_a = train_test_split(X_small, y_small, random_state=0)
model_small = LogisticRegression(max_iter=2000).fit(X_small_a, y_small_a)


#### -------------Validation model
## X & y
model_logreg_training = model_logreg.score(X_train, y_train)
model_logreg_val = model_logreg.score(X_val, y_val)
print(f"Training set score for X & y: {model_logreg_training:.3f}")
print(f"Validation set score for X & y: {model_logreg_val:.3f}")
print()

# Only first two columns of X and y
model_partial_training = model_partial.score(X_train_partial, y_train_partial)
model_partial_val = model_partial.score(X_val_partial, y_val_partial)
print(f"Training set score for partial X & y: {model_partial_training:.3f}")
print(f"Validation set score for partial X & y: {model_partial_val:.3f}")
print()

# X_small and y_small
model_small_training = model_small.score(X_small_a, y_small_a)
model_small_val = model_small.score(X_val_a, y_val_a)
print(f"Training set score for small X & y: {model_small_training:.3f}")
print(f"Validation set score for small X & y: {model_small_val:.3f}")
print()



##----------------

results = pd.DataFrame(columns = ["Data Size", "Training accuracy", "Validation accuracy"])

results.loc[0] = [X.size, model_logreg_training, model_logreg_val]
results.loc[1] = [X.iloc[:,:2].size, model_partial_training, model_partial_val]
results.loc[2] = [X_small.size, model_small_training,model_small_val]

print(results.round(3))


Training set score for X & y: 0.928
Validation set score for X & y: 0.938

Training set score for partial X & y: 0.608
Validation set score for partial X & y: 0.613

Training set score for small X & y: 0.936
Validation set score for small X & y: 0.931

   Data Size  Training accuracy  Validation accuracy
0   262200.0              0.928                0.938
1     9200.0              0.608                0.613
2    13110.0              0.936                0.931


### Questions (4 marks)
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

Answer 1:

new : When we use the full data sets,with all the features we get training accuracy score of 0.928 and validation score of 0.938. Since they have high scores and close to each other, we can say that the model performed well.

As you use less data(ie: parital columns used for X), the training score was 0.608 and validation score was 0.613. From this low scores for both, we see that less features is used to train and this models becomes too simple and underfits. 

For the small sample size, it has training score of 0.936 and validation score of 0.931, however it has been trained on less data, therefore it could of memorirized it and thus leading to higher training and validation accuracy.

-----
Answer 2:
In this instance false negative is when you get the spam email in your inbox, when it should of been flagged as spam and go to the spam folder.
False positive is when a non-spam email is flagged as an spam and it went to the spam folder.

False positive is worse in my opinion, because it can be a problem when an imporant email gets flagged as a spam.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

Answer 1:
I used the previous examples we done in class, such as logestic regression, and modified the code as per the question requirements. I also took assistance for the generative AI called ChatGPT. 

Answer 2:
I completed the steps as per the instrustions on this file. I started from Step 0. This was a good way to tackle the problem step by step to understand what is going on.

Answer 3:
The prompts I used in ChatGPT are "How did you format the signitficant digits in a f string". I did not need to modify the code becauase the formatting you have to do it in a certain way. 

Link: https://chat.openai.com/

Answer 4:
Initally, I had challenge starting the problem, but I looked at previous examples done on class as reference, and how they implemented the code. It was challenging, but I understood there was a pattern in machine learning model and understanding how to split the data.

## Part 2: Regression (10.5 marks total)

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [8]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y

from yellowbrick.datasets import load_concrete
X,y = load_concrete()

print(X.shape)
print(y.shape)

print(type(X))
print(type(y))

(1030, 8)
(1030,)
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>


### Step 2: Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [9]:
# TO DO: Check if there are any missing values and fill them in if necessary
missing_values_X = X.isnull().sum().sum()
missing_values_y = y.isnull().sum().sum()

print(missing_values_X)
print(missing_values_y)

0
0


### Step 3: Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [10]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0
from sklearn.linear_model import LinearRegression
model = LinearRegression()

X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
lr = LinearRegression().fit(X_train, y_train)




### Step 4: Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [11]:
# TO DO: ADD YOUR CODE HERE

from sklearn.metrics import mean_squared_error, r2_score

y_train_pred = lr.predict(X_train)
y_val_pred = lr.predict(X_val)

mse_train = mean_squared_error(y_train, y_train_pred)
mse_val = mean_squared_error(y_val, y_val_pred)


print(f"Training set score for Mean Squared Error: {mse_train:.2f}")
print(f"Validation set score for Mean Squared Error: {mse_val:.2f}")
print()
#-----

r2_train = r2_score(y_train, y_train_pred)
r2_val = r2_score(y_val, y_val_pred)


print(f"Training set score for R2 Score: {r2_train:.3f}")
print(f"Validation set score for R2 Score: {r2_val:.3f}")

Training set score for Mean Squared Error: 111.36
Validation set score for Mean Squared Error: 95.90

Training set score for R2 Score: 0.611
Validation set score for R2 Score: 0.623


### Step 5: Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [12]:
# TO DO: ADD YOUR CODE HERE
results = pd.DataFrame(columns=["Training accuracy", "Validation accuracy"], index=["Mean Squared Error", "R2 Score"])
results["Training accuracy"] = [mse_train, r2_train]
results["Validation accuracy"] = [mse_val, r2_val]

print(results.round(3))



                    Training accuracy  Validation accuracy
Mean Squared Error            111.358               95.904
R2 Score                        0.611                0.623


### Questions (2 marks)
1. Did using a linear model produce good results for this dataset? Why or why not?

Answer 1: In the linaear model, the Mean Squared Error(MSE) training score is  111.358, Validation score is 95.904. And for R2 Score training score is 0.611 and Validation score is 0.623.


Also R2 scores is way less than 1, where values closer to 1 has a better fit. Both the Training and validation scores are about 60%, therefore the model explains 60% of the variance in the data. This also shows that the models underfits and it is too simple.

Therefore, I believe Linear model, did not produce good results.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

Answer 1:
I used the previous examples we done in class like i mentioned before, such as Linear Regression, and modified the code as per the question requirements. I also took assistance for the generative AI called ChatGPT. 

Answer 2:
I completed the steps as per the instrustions on this file. I started from Step 1. This was a good way to tackle the problem step by step to understand what is going on.

Answer 3:
The prompts I used in ChatGPT are "how do you import r2 and mse libraries from sklearn and to calculate the mse and r2 scores". I did not need to modify the code becauase the formatting you have to do it in a certain way in order to import these libaries. And to calculate the r2 and mse scores, it has to be done from a function call and that you have to do it in a certain way. 

Link: https://chat.openai.com/

Answer 4:
Initally, I had challenge of calculating the mse and r2 scores becuase i could not find it from the examples when I looked for reference, but ChatGPT was able to assit me in importing those libraries and how they are clauclated. This has helped me to complete the calculation for mse and r2 scores.


## Part 3: Observations/Interpretation (3 marks)

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


In part 1, we can see that when we use only the first two columns, we get the training and validiation scores around 0.6. Since is way less than one it is less accurate model compareded to the full data set and the small data set. This means when you use less data/less features used to train, this models becomes too simple and underfits. 

For Full data set in part 1, has training score of 0.928 and vaildation score of 0.938. This has slightly worse results than small data (Training score: 0.936 and Validation score:0.931). The small data has been only trained on smaller set of data, so the scores received was higher than original data set for training and validation scores since it has been trained on less data. Therefore, I see that when you give less data to train, it most likely memorizes the data, therefore giving a higher training and validation scores.

In Part 2, we see a training score of 0.611 and validation score of 0.623, this shows that it is underfitting in R2 scores. This suggests that the model is not fitting well, and not performing better on the validation data. This could mean high bias and the model may be too simple.



## Part 4: Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


What I liked about this assignment is that I got a hands on experience to build linear and logestic models. I got to see how these model works and a good learning experience.

I found it confusing at first on why we had to do train split on three different data sets for Question 1. But later i understood why we had to do the train split.

## Part 5: Bonus Question (4 marks)

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [13]:
# TO DO: ADD YOUR CODE HERE
from sklearn.linear_model import Ridge, Lasso
from sklearn.metrics import mean_squared_error, r2_score

from yellowbrick.datasets import load_concrete
X,y = load_concrete()


##Ridge regression
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
ridge_model = Ridge(alpha= 1).fit(X_train, y_train)

y_train_pred = ridge_model.predict(X_train)
y_val_pred = ridge_model.predict(X_val)

mse_train = mean_squared_error(y_train,y_train_pred)
mse_val = mean_squared_error(y_val,y_val_pred)

print(f"Ridge Regression - Training set score for Mean Squared Error: {mse_train:.2f}")
print(f"Ridge Regression - Validation set score for Mean Squared Error: {mse_val:.2f}")


r2_train = r2_score(y_train, y_train_pred)
r2_val = r2_score(y_val, y_val_pred)
print(f"Ridge Regression - Training set score for R2 Score: {r2_train:.3f}")
print(f"Ridge Regression - Validation set score for R2 Score: {r2_val:.3f}")
print()

## Lasso
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
lasso_model = Lasso(alpha= 1).fit(X_train, y_train)

y_train_pred = lasso_model.predict(X_train)
y_val_pred = lasso_model.predict(X_val)

mse_train = mean_squared_error(y_train,y_train_pred)
mse_val = mean_squared_error(y_val,y_val_pred)

print(f"Lasso Regression - Training set score for Mean Squared Error: {mse_train:.2f}")
print(f"Lasso Regression - Validation set score for Mean Squared Error: {mse_val:.2f}")

r2_train = r2_score(y_train, y_train_pred)
r2_val = r2_score(y_val, y_val_pred)
print(f"Lasso Regression - Training set score for R2 Score: {r2_train:.3f}")
print(f"Lasso Regression - Validation set score for R2 Score: {r2_val:.3f}")
print()






Ridge Regression - Training set score for Mean Squared Error: 111.36
Ridge Regression - Validation set score for Mean Squared Error: 95.90
Ridge Regression - Training set score for R2 Score: 0.611
Ridge Regression - Validation set score for R2 Score: 0.623

Lasso Regression - Training set score for Mean Squared Error: 111.42
Lasso Regression - Validation set score for Mean Squared Error: 95.58
Lasso Regression - Training set score for R2 Score: 0.611
Lasso Regression - Validation set score for R2 Score: 0.625



I got 0.611 for training for both lasso and ridge regression and 0.623 for Ridge regression and 0.625 for Lasso regression.

I tried different numbers of alpha numbers, but Alpha number of 1 has the best R2 Square becuause it gave me the highest R2 scores and the one closest to 1. 
However, the score is not good enough becuase it is not close to 1 and it is further away. 

It is difficult to say which method performed better becuase all the methods performed gave similar results about 0.611 for training for ridge,lasso, and linear and 0.623 for Ridge regression and linear, and 0.625 for Lasso regression., therefore all the methods used was not useful at all.

Since the R2 are low, it indicates the model is underfitting and the objective of lasso and ridge is to regularize the model and reduce overfitting, however, in this case it is not needed and we should try a more complex model.

Therefore, I believe the score is  not good enough.